### ISSUE TITLE
source-repository-package branch failure

### ISSUE THREAD

**Issue by georgefst** [haskell/cabal#6888]
With:

```
source-repository-package
  type: git
  location: git://github.com/georgefst/pretty-simple.git
  branch: experimental
```

at the top of my `cabal.project`, I get the failure `fatal: '--detach' cannot be used with '-b/-B/--orphan'`. Unclear what's causing this. 

`git` version is `2.7.4`, which is pretty old (can't upgrade right now on my work machine), so if that's the issue, I'll just work around it.

**Comment by phadej** [comment]
An example output

```
Project settings changed, reconfiguring...
creating /code/public-haskell/tree-diff/dist-newstyle
creating /code/public-haskell/tree-diff/dist-newstyle/cache
/usr/bin/git --version
creating /code/public-haskell/tree-diff/dist-newstyle/src
/usr/bin/git clone --no-checkout 'git://github.com/georgefst/pretty-simple.git' /code/public-haskell/tree-diff/dist-newstyle/src/pretty-si_-3997a25a56f97597
Cloning into '/code/public-haskell/tree-diff/dist-newstyle/src/pretty-si_-3997a25a56f97597'...
remote: Enumerating objects: 55, done.
remote: Counting objects: 100% (55/55), done.
remote: Compressing objects: 100% (39/39), done.
remote: Total 1204 (delta 16), reused 44 (delta 11), pack-reused 1149
Receiving objects: 100% (1204/1204), 288.14 KiB | 1.56 MiB/s, done.
Resolving deltas: 100% (514/514), done.
/usr/bin/git checkout --detach --force experimental --
fatal: '--detach' cannot be used with '-b/-B/--orphan'
```

and looks like we have an unsolved TODO

https://github.com/haskell/cabal/blob/5ce6b08b332fecf4e5100595bbe55edf6060f99c/cabal-install/Distribution/Client/VCS.hs#L628-L629

**Comment by liftM** [comment]
For anyone else who arrives at this, you can work around it by `cd`ing into the cloned directory in `dist-newstyle` and manually running the correct `git checkout` yourself.

**Comment by georgefst** [comment]
After all this time, I have suddenly accidentally discovered that `branch` is optional, and if you're specifying `tag` anyway, fairly pointless. Now that I think about it, it would be weird if this wasn't the case.

Admittedly, now that I look, the documentation clearly states:
> This field is optional.

So this issue isn't quite as high priority as I'd thought. Would still be nice though!

**Comment by tomjaguarpaw** [comment]
I also came across this. On searching for the error `fatal: '--detach' cannot be used with '-b/-B/--orphan'` I arrived here. To make @georgefst's workaround crystal clear: don't use `branch` just use `tag` (which actually means to seem "commit"). For example, in `cabal.project`:

```cabal
packages: *

source-repository-package
 type: git
 location: git@github.com:purebred-mua/purebred-email.git
 tag: 19e06063d03c477f33f3add82c1b4f7c6e77b801
```

**Comment by georgefst** [comment]
> just use `tag` (which actually means to seem "commit").

I think "tag" is an attempt to use VCS-agnostic terminology. But it is a little confusing since Git has its own separate concept of a "tag".

**Comment by fgaz** [comment]
maybe "revision" would be a more appropriate name (though changing it is probably more pain than it's worth)

**Comment by georgefst** [comment]
> though changing it is probably more pain than it's worth

Strongly agreed!

**Comment by tomjaguarpaw** [comment]
Personally I don't agree. `tag` is sufficiently misleading in the context of git that I think it would be better to provide a less misleading alternative. Perhaps introduce `revision` or `rev` and deprecate `tag`? (Although if `tag` is used for genuine git tags then it shouldn't be deprecated for that purpose.)

**Comment by georgefst** [comment]
Fair enough, I can see the argument. This is getting very off-topic though.

**Comment by purefn** [comment]
> After all this time, I have suddenly accidentally discovered that branch is optional, and if you're specifying tag anyway, fairly pointless. Now that I think about it, it would be weird if this wasn't the case.

This wouldn't be weird at all. I would prefer to specify both branch and tag (aka revision), if it meant that cabal would only fetch that specific branch and the changes in it. Most of the time this is the behavior I want, I don't want the whole repository.

**Comment by tomjaguarpaw** [comment]
> I would prefer to specify both branch and tag (aka revision), if it meant that cabal would only fetch that specific branch and the changes in it. Most of the time this is the behavior I want, I don't want the whole repository.

Specifying the branch has no relation to how much of the repository cabal (or rather git run by cabal) will fetch.

**Comment by purefn** [comment]
> Specifying the branch has no relation to how much of the repository cabal (or rather git run by cabal) will fetch.

I'm aware of that, but it *should* is what I was trying to get at.

**Comment by tomjaguarpaw** [comment]
Yes indeed that would be a nice feature. The reason I said that I don't see the connection was that I believed the same feature could be used to shallow-fetch commit hashes too, not just branches or tags. However, [that functionality seems to have been removed](https://stackoverflow.com/questions/3489173/how-to-clone-git-repository-with-specific-revision-changeset/3489576#3489576) at some point during git's evolution, sadly.

This is rather frustrating since fetching a branch name (or even a tag) is not reproducible, but fetching a commit hash is. Therefore we would ideally want to shallow-fetch commit hashes!

**Comment by tomjaguarpaw** [comment]
Having said that, someone has, in the last four months, created [a go program to do shallow fetches by sha](https://github.com/robherley/shallow-fetch-sha#shallow-fetch-sha-sfs-%EF%B8%8F) so this recipe must work with at least *some* git servers. Theoretically cabal could try to shallow fetch first and then do a full fetch if shallow doesn't work (but I don't wish that maintenance burden on the cabal developers).

**Comment by tomjaguarpaw** [comment]
Update: it is possible to shallow-fetch commit hashes, as long as the server enables support for it, which GitHub does, for example, and as long as you use the full, unabbreviated commit hash. See my edits above.

**Comment by purefn** [comment]
Yes, you can do that as long as the commit is on the default branch. As far as I'm aware, if it is not on the default branch, then it will not work.

**Comment by tomjaguarpaw** [comment]
Here's an example of fetching from a commit that is on `freeze-docs` and not on the default branch (`master`):

```
% git init && git remote add origin https://github.com/haskell/cabal.git
% git fetch --depth 1 origin c4230704d739115a13284d8f713b8ef781309d5b
remote: Enumerating objects: 3266, done.
remote: Counting objects: 100% (3266/3266), done.
remote: Compressing objects: 100% (2518/2518), done.
remote: Total 3266 (delta 480), reused 1998 (delta 299), pack-reused 0
Receiving objects: 100% (3266/3266), 2.34 MiB | 8.78 MiB/s, done.
Resolving deltas: 100% (480/480), done.
From https://github.com/haskell/cabal
 * branch            c4230704d739115a13284d8f713b8ef781309d5b -> FETCH_HEAD
```

**Comment by tomjaguarpaw** [comment]
(Admittedly you can only fetch commits that are on *some* branch, but that's a much less severe requirement.)

**Comment by purefn** [comment]
Holy cow! I wish all tools did it this way! The amount of time I've spent waiting for both stack and cabal to do full clones of git repos can literally be counted in days. If they did this that would probably cut it down to less than an hour.

**Comment by tomjaguarpaw** [comment]
Yeah, this feature seems desirable and simple enough that I'm willing to proclaim I _want_ to implement it (but then probably not actually implement it).

**Comment by jneira** [comment]
hi, after https://github.com/haskell/cabal/pull/7625 the code strated to use `git reset` instead `git checkout` and now the error is different, with this `cabal.project`:

```
packages: .

source-repository-package
  type: git
  location: https://github.com/georgefst/pretty-simple.git
  branch: web
```

i got:

```
Running: "D:\dev\app\git\cmd\git.exe" "clone" "--no-checkout" "https://github.com/georgefst/pretty-simple.git" "D:\dev\ws\haskell\issues\6888\dist-newstyle\src\pretty-si_-592d60f12a0e425"
Running: "D:\dev\app\git\cmd\git.exe" "submodule" "deinit" "--force" "--all"
Running: "D:\dev\app\git\cmd\git.exe" "reset" "--hard" "web" "--"
fatal: Failed to resolve 'web' as a valid revision.
```

I think the underlying problem is the same though but i wanted to update the info

**Comment by noinia** [comment]
I finally got sufficiently annoyed by this issue that I decided to fix it..... (at least for git repos)

**Comment by soulomoon** [comment]
Same here

**Comment by isovector** [comment]
What's the status on this? It's a pretty rough bug.
